### PR TITLE
Extended S3 URI format

### LINKED ISSUE #67519: S3 URI does not handle endpoints not start with s3/oss
(you don't have to strictly follow this form)

**Company or project name**
Alibaba cloud
Put your company name or project description here

**Describe the unexpected behaviour**

For example, with the endpoint like:
`https://bucket-test.cn-beijing-internal.oss-data-acc.aliyuncs.com/cc-2zeh496zqm0g6e09g/`

The expected parsing result is:
bucket ： bucket-test
url: cn-beijing-internal.oss-data-acc.aliyuncs.com
path: cc-2zeh496zqm0g6e09g

However, the the bucket would be parsed into "bucket-test.cn-beijing-internal".

There are two problems here
1.  The endpoint is split according to ".s3/.oss" as the delimiter and is not compatible with situations that do not start with s3. The more appropriate approach would be to use the first '.' to do the split.

2.  Unable to specify whether to use virtual host or path style.  Even we configured as second_path format. The endpoint still matches the virutal_host's regular rule. 

I can provide a PR if necessary. But before that, some input is needed on the direction of problem solving.

### PR DESCRIPTION
**Pull request by baolinhuang**
### Changelog category (leave one):
- New Feature
- Experimental Feature
- Improvement
- Performance Improvement
- Backward Incompatible Change
- Build/Testing/Packaging Improvement
- Documentation (changelog entry is not required)
- Critical Bug Fix (crash, LOGICAL_ERROR, data loss, RBAC)
- Bug Fix (user-visible misbehavior in an official stable release)
- CI Fix or Improvement (changelog entry is not required)
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

close #67519

### CODE REVIEW

**Comment by nickitat** [comment]
looking at the uri provided in the issue i don't see strong reasons why simply adjusting `virtual_hosted_style_pattern` maybe with some minor code changes wouldn't be enough. have you explored that road?
there is a test for this code (`src/IO/tests/gtest_s3_uri.cpp`) please don't forget to extend it.

**Comment by baolinhuang** [comment]
> looking at the uri provided in the issue i don't see strong reasons why simply adjusting `virtual_hosted_style_pattern` maybe with some minor code changes wouldn't be enough. have you explored that road? there is a test for this code (`src/IO/tests/gtest_s3_uri.cpp`) please don't forget to extend it.


1. It is indeed difficult to adapt the Issue URL format through minor modifications.
	The first half of the URL: cn-beijing-internal.oss-data-acc.aliyuncs.com (cn-beijing-internal) is the OSS region information, and there is no unified format.
2. Regarding the current implementation of S3 URI, I think it is unreasonable. There are two main problems.
	a. Use .s3 for regular distinction instead of ., which causes the problems encountered in the current issue. The recently submitted https://github.com/ClickHouse/ClickHouse/commit/0ad3ed5 also deals with similar problems. This regular method will become increasingly difficult to be compatible with all situations.
	b. Unable to explicitly specify the url format, virtual host or path
3. I tested other databases, and in the case of this issue, both doris and presto can parse the bucket correctly.
	Among them, doris uses '.' for bucket interception, and supports parameters to specify the style of the URL.

To sum up, I just proposed the solution in PR. However, I also know that this approach will lead to more modifications.
If there is a way to solve this issue with minor modifications, I would be happy to accept it.

**Comment by baolinhuang** [comment]
If you think the above patch modification is a bit large.

I am glad to provide a small patch for the scenario we are currently facing, and adapt the domain name (which has oss-data-acc) individually. Does the same thing as aws private link interface.

**Comment by nickitat** [comment]
pls check https://github.com/ClickHouse/ClickHouse/pull/69400 if it does the job in your case?

**Comment by baolinhuang** [comment]
> pls check #69400 if it does the job in your case?

Yes. It works.

**Comment by nickitat** [comment]
ok, closing this one in favor of https://github.com/ClickHouse/ClickHouse/pull/69400
